### ISSUE TITLE
[rocksdb] Investigate compression scheme for RocksDB

### ISSUE THREAD

**Issue by agourlay** [qdrant/qdrant#804]
**Is your feature request related to a problem? Please describe.**
A quick profiling session on a heavy write workload shows that Snappy can take up to 2% of CPU time.
![snappy](https://user-images.githubusercontent.com/606963/178513624-622407d2-c69f-49d8-83e8-58237b5a3a02.png)

The official doc says:

```
Use options.compression to specify the compression to use. 
By default it is Snappy. We believe LZ4 is almost always better than Snappy.
```
https://github.com/facebook/rocksdb/wiki/Compression

**Describe the solution you'd like**
Investigate if it can be beneficial to switch to another compression scheme.

e.g.

```rust
options.set_compression_type(DBCompressionType::Lz4);
```

**Comment by generall** [comment]
Criteria for this issue:

- It is possible to make switch to a new compression backward compatible
- How well it compresses datasets with/without payload: size, write/read speed benches

**Comment by thomas-k-cameron** [comment]
I have done something similar in the past.

Here is my take:
- zstd should be faster based on some benchmarks
- Also, the official [wiki](https://github.com/facebook/rocksdb/wiki/Compression) recommends zstd

Let me try it.

**Comment by thomas-k-cameron** [comment]
I bechmarked it.

Here is a table of results.

I used the benchmarking script in benches directory. 

I'm don't know how scientific it is but I think it gives you a rough idea.

To summarize, upsert performance is `zstd` is on par with `snappy` but `zstd` makes it significantly faster when it comes to search performance.

| compression | benchmark          | grpc_req_duration | iteration duration | iterations   |
| ----------- | ------------------ | ----------------- | ------------------ | ------------ |
| snappy      | run_grpc_upsert.sh | avg=21.41ms       | avg=63.29ms        | 449.381818/s |
| zstd        | run_grpc_upsert.sh | avg=20.73ms       | avg=62.6ms         | 427.646846/s |
| snappy      | run_grpc_search.sh | avg=38.5ms        | avg=60.26ms        | 248.779121/s |
| zstd        | run_grpc_search.sh | avg=27.82ms       | avg=45.5ms         | 329.470261/s |

If you want to be more adventurous, I think you can consider adopting something else like this one.
https://github.com/bytedance/terarkdb

**Comment by thomas-k-cameron** [comment]
So I did some benchmarking using scripts I found on `bench/services`. 
It turns out that Lz4, snappy and zstd makes almost no difference both in benchmark performance and storage usage.

My guess is that dataset is either too small or nature of the dataset makes it difficult to make it well-compressed.

I think we might have a better result if we `tune` them better or have a lot of data, but I'm not quite sure if the effort is going to worth it.

You can find the source code here.
https://github.com/thomas-k-cameron/qdrant/tree/zstd

You can find the benchmark results here.
https://github.com/thomas-k-cameron/qdrant/tree/zstd/benchmark-result

**Comment by thomas-k-cameron** [comment]
So I tried to benchmark it with `cargo bench`. 

I tried every single one of the bench mark target that I could find on `lib/segment/Cargo.toml`, but I always end up with division error. 

Is there anything that I'm missing?

e.g. 
```
@thomas-k-cameron ➜ /workspaces/qdrant/lib/segment (bb5e2402) $ RUSTFLAGS="--cfg zstd" cargo bench struct_boolean_query_points
    Finished bench [optimized] target(s) in 0.22s
     Running unittests src/lib.rs (/workspaces/qdrant/target/release/deps/segment-346d2e5a8184dbc4)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 264 filtered out; finished in 0.00s

     Running benches/boolean_filtering.rs (/workspaces/qdrant/target/release/deps/boolean_filtering-d82aae13e91e07f3)
Gnuplot not found, using plotters backend
thread 'main' panicked at lib/segment/benches/boolean_filtering.rs:48:9:
attempt to divide by zero
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: bench failed, to rerun pass `--bench boolean_filtering`
```

**Comment by generall** [comment]
Practically all current benchmarks we have does not include RocksDB in the process. I can see 2 relatively simple cases which we can try to compare performance:

- Storage size of the dataset with payload
- Collection loading speed after restart

For both estimations you would need a relatively big dataset with payload. 

I suggest using something like this: https://storage.googleapis.com/common-datasets-snapshots/arxiv_abstracts-3083016565637815127-2023-06-02-07-26-29.snapshot

**Comment by thomas-k-cameron** [comment]
Thanks for the tip. Let me try it out.

**Comment by agourlay** [comment]
@thomas-k-cameron Thank you for looking into this :+1: 

> I tried every single one of the bench mark target that I could find on lib/segment/Cargo.toml, but I always end up with division error.

The command `cargo bench` takes as an argument the name of the benchmark as defined in the `Cargo.toml` and not the name of the benchmark function.

In your case it would be `cargo bench boolean_filtering`

Not sure how to run a single benchmark function.

Good luck with the snapshot based investigation!

**Comment by thomas-k-cameron** [comment]
Hey @agourlay 

Thanks, it worked.
Yes, i totally forgot to put `--bench` in there 😢

**Comment by thomas-k-cameron** [comment]
Hey @generall, @agourlay 

I measured the time required to start up the server and I was able to observe a noticeable difference this time.
- Time required to start the server

It took 4 seconds on `lz4`, 3 seconds on `zstd` and 10 seconds on `snappy`, so I believe zstd is the fastest here.

- Storage size

I couldn't observe any noticeable difference here; I measured it by `du storage -h -d 1`, though I'm guessing that this doesn't properly account for other things like the WAL file.

I'm guessing that it is not a bad idea to set `zstd` as a default compression. How do you think?

**Comment by timvisee** [comment]
Sounds good! Thanks for giving it another proper test. Do you have this available on a branch somewhere, so I can give it a quick spin too?

It would be that `snappy` is a good default for most data, but not for what we put into it. I wonder if it would make sense for us to use different compression schemes for vectors versus payload (if even possible).

**Comment by thomas-k-cameron** [comment]
So here is the test result. It's a copy and paste of my shell.

https://github.com/thomas-k-cameron/qdrant/blob/zstd/examples.md

You can change the compression algorithm by changing the env variable e.g. `ROCKSDB_COMPRESSION="lz4" cargo run --release`

**Comment by agourlay** [comment]
Great work!

Any idea how the migration path would look like to switch the compression scheme on existing databases?

**Comment by thomas-k-cameron** [comment]
I think the easiest would be to go through a snapshot.
e.g. Take a snapshot of the storage and recover the snapshot the next time you start up the database.

In this case, you just need to give users the option to switch between different compression algorithm; This will make the development fairly simple.

The fancier way would be to setup a dedicated API endpoint that allows users to switch between different algorithm, though I don't think this is going to be technically complicated and bring potentially confusing user experience.

As far as I'm concerned, when you try to open rocksdb file compressed with different algorithm, it is not going to work; it panic-ed when I tried it.

If you wouldn't mind, I can create a PR and discuss about the implementation there.

**Comment by thomas-k-cameron** [comment]
@agourlay How is it going?

**Comment by agourlay** [comment]
@thomas-k-cameron sorry this fell off my radar :)

Ideally we do not want users to care about the underlying RocksDB implementation.
I'd rather not expose configuration knobs specific to our dependencies.

What about a different approach where we use `zstd` only on new collection and keep supporting both until `snappy` slowly fade away?

**Comment by dacox** [comment]
@agourlay @thomas-k-cameron 

i dont have anything to add here, just trying to hunt down serious memory issues we are having, seemingly related to payloads.

Not sure if an optimization like this would help or not - https://github.com/orgs/qdrant/discussions/3501#discussioncomment-8383986

**Comment by thomas-k-cameron** [comment]
> @agourlay @thomas-k-cameron
> 
> i dont have anything to add here, just trying to hunt down serious memory issues we are having, seemingly related to payloads.
> 
> Not sure if an optimization like this would help or not - https://github.com/orgs/qdrant/discussions/3501#discussioncomment-8383986

It can help but it's not going to fix anything.

Data types available on Rust std library allocates more memory than it needs when it doesn't have any additional capacity.
e.g. 
```rust
let mut vector = vec![1i64, 2i64];
assert_eq!(vector.capacity() == 2);
vector.push(3i64);
assert_eq!(vector.capacity() == 4);
```

It can add up really really quickly if you have a lot of elements.  
Many data types comes with `shrink_to_fit()`, which frees any execs allocation.   
Another thing you could do is `into_boxed_slice`, boxed slice doesn't allocate any extra memory but you can't add anything to it.  

As far as I'm concerned, qdrant doesn't shrink or use functions for avoiding exec allocation.

**Comment by timvisee** [comment]
> qdrant doesn't shrink or use functions for avoiding exec allocation.

We are aware of capacity allocations in our code base. In many where it makes sense we control this and limit growing.

But, we don't use `shrink_to_fit()` very often because this can be very inefficient. Instead you should prevent growing your capacity when it's not needed.

In many data structures we still want mutability to add new entries at any time. If you'd always set exact capacity here, with `shrink_to_fit()` for example, you always need to reallocate for every insertion which is very costly.

> Data types available on Rust std library allocates more memory than it needs when it doesn't have any additional capacity.

So, this is for good reason.

A lot of data types we have that are not mutable, already have their exact capacity set.

**Comment by thomas-k-cameron** [comment]
Thanks for the context!

**Comment by thomas-k-cameron** [comment]
> @thomas-k-cameron sorry this fell off my radar :)
> 
> Ideally we do not want users to care about the underlying RocksDB implementation. I'd rather not expose configuration knobs specific to our dependencies.
> 
> What about a different approach where we use `zstd` only on new collection and keep supporting both until `snappy` slowly fades away?

Sounds like a great idea.

**Comment by agourlay** [comment]
We are enabling first LZ4 based on some internal benchmarks https://github.com/qdrant/qdrant/pull/4723

@thomas-k-cameron Thank you very much for the initial work on this :clap:

**Comment by thomas-k-cameron** [comment]
Hi @agourlay 
I'm very sorry that I couldn't pull through with it.
I'm very happy that my work helped you!

**Comment by agourlay** [comment]
released in [1.11.0](https://github.com/qdrant/qdrant/releases/tag/v1.11.0)
